### PR TITLE
*: implement auto analyze

### PR DESCRIPTION
**Pull request by lamxTyler**
This PR implements the auto analyze.

- It uses the `OwnerManager` to choose the stats owner, and only the stats owner can do auto analyze.

- It will only auto analyze for the newly created table and index. For the table, only after `20*statsLease` with no insert, the analyze operation will be triggered. For the index, we will analyze immediately after we know it's existence.

- To reduce the impact on the system, it will only analyze only one table or index in one time. Also, the priority of the `analyze` is set to low priority.

PTAL @hanfei1991 @winoros

### CODE REVIEW

**Comment by lamxTyler** [comment]
Currently, if the `statsLease` is 0, there will be no auto analyze. Wondering if it needs a separate parameter to control it.

File: ddl/ddl.go [hunk 1]
```diff
@@ -273,7 +273,7 @@ func newDDL(ctx goctx.Context, etcdCli *clientv3.Client, store kv.Storage,
 		manager = NewMockOwnerManager(id, cancelFunc)
 		syncer = NewMockSchemaSyncer()
 	} else {
-		manager = NewOwnerManager(etcdCli, id, cancelFunc)
```
Comment (zimulala) [inline ddl/ddl.go]: Add `ddl` as a constant.

File: ddl/owner_manager.go [hunk 2]
```diff
@@ -57,37 +57,41 @@ const (
...
+	id      string // id is the ID of the manager.
+	key     string
+	prompt  string
+	etcdCli *clientv3.Client
+	cancel  goctx.CancelFunc
 }
 
 // NewOwnerManager creates a new OwnerManager.
-func NewOwnerManager(etcdCli *clientv3.Client, id string, cancel goctx.CancelFunc) OwnerManager {
+func NewOwnerManager(etcdCli *clientv3.Client, prompt string, id string, key string, cancel goctx.CancelFunc) OwnerManager {
```
Comment (zimulala) [inline ddl/owner_manager.go]: s/prompt string, id string, key string/prompt, id, key string

File: domain/domain.go [hunk 3]
```diff
@@ -575,6 +575,26 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
...
 		}
 	}(do)
+	go func(do *Domain) {
+		id := do.ddl.OwnerManager().ID()
+		cancelCtx, cancelFunc := goctx.WithCancel(goctx.Background())
+		var statsOwner ddl.OwnerManager
+		if do.etcdClient == nil {
+			statsOwner = ddl.NewMockOwnerManager(id, cancelFunc)
+		} else {
+			statsOwner = ddl.NewOwnerManager(do.etcdClient, "stats", id, statistics.StatsOwnerKey, cancelFunc)
```
Comment (zimulala) [inline domain/domain.go]: The same as `stats`.

File: domain/domain.go [hunk 4]
```diff
@@ -575,6 +575,26 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
...
+		}
+		statsOwner.CampaignOwners(cancelCtx)
```
Comment (zimulala) [inline domain/domain.go]: Do we need to check the error?

File: domain/domain.go [hunk 5]
```diff
@@ -575,6 +575,26 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
...
+		for {
+			if statsOwner.IsOwner() {
+				err := do.statsHandle.HandleAutoAnalyze(do.InfoSchema())
+				if err != nil {
+					log.Error(errors.ErrorStack(err))
```
Comment (zimulala) [inline domain/domain.go]: Complete the error message. Add the package and function information. You can refer to other logs.

**Comment by hanfei1991** [comment]
@lamxTyler Can we write a `owner` pakcage independent of `ddl` package ?

File: ddl/owner_manager.go [hunk 6]
```diff
@@ -134,25 +139,25 @@ func newSession(ctx goctx.Context, flag string, etcdCli *clientv3.Client, retryC
...
+func (m *ownerManager) campaignLoop(ctx goctx.Context, etcdSession *concurrency.Session) {
+	idInfo := fmt.Sprintf("%s ownerManager %s", m.key, m.id)
 	var err error
 	for {
 		select {
 		case <-etcdSession.Done():
-			log.Infof("[ddl] %s etcd session is done, creates a new one", idInfo)
-			etcdSession, err = newSession(ctx, idInfo, m.etcdCli, newSessionRetryUnlimited, ManagerSessionTTL)
+			log.Infof("[%s] %s etcd session is done, creates a new one", m.prompt, idInfo)
+			etcdSession, err = newSession(ctx, m.prompt, idInfo, m.etcdCli, newSessionRetryUnlimited, ManagerSessionTTL)
```
Comment (zimulala) [inline ddl/owner_manager.go]: Can we use a log prefix instead of `m.prompt` and `idInfo` here?  We can pass an argument.

File: ddl/owner_manager.go [hunk 7]
```diff
@@ -260,10 +259,3 @@ func (m *ownerManager) watchOwner(ctx goctx.Context, etcdSession *concurrency.Se
 		}
 	}
 }
-
-func init() {
-	err := setManagerSessionTTL()
```
Comment (zimulala) [inline ddl/owner_manager.go]: Why remove this ？

File: ddl/syncer.go [hunk 8]
```diff
@@ -123,7 +123,7 @@ func (s *schemaVersionSyncer) Init(ctx goctx.Context) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
-	s.session, err = newSession(ctx, s.selfSchemaVerPath, s.etcdCli,
+	s.session, err = newSession(ctx, "ddl", s.selfSchemaVerPath, s.etcdCli,
```
Comment (zimulala) [inline ddl/syncer.go]: Use `ddlPrompt` directly.

File: ddl/syncer.go [hunk 9]
```diff
@@ -140,7 +140,7 @@ func (s *schemaVersionSyncer) Done() <-chan struct{} {
 
 // Restart implements SchemaSyncer.Restart interface.
 func (s *schemaVersionSyncer) Restart(ctx goctx.Context) error {
-	session, err := newSession(ctx, s.selfSchemaVerPath, s.etcdCli,
+	session, err := newSession(ctx, "ddl", s.selfSchemaVerPath, s.etcdCli,
```
Comment (zimulala) [inline ddl/syncer.go]: Ditto.

File: domain/domain.go [hunk 10]
```diff
@@ -575,6 +575,29 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
 			}
 		}
 	}(do)
+	go func(do *Domain) {
```
Comment (zimulala) [inline domain/domain.go]: This function is too long. Can we extract this logic into a function?

File: domain/domain.go [hunk 11]
```diff
@@ -575,6 +575,29 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
...
+		id := do.ddl.OwnerManager().ID()
+		cancelCtx, cancelFunc := goctx.WithCancel(goctx.Background())
+		var statsOwner ddl.OwnerManager
+		if do.etcdClient == nil {
+			statsOwner = ddl.NewMockOwnerManager(id, cancelFunc)
+		} else {
+			statsOwner = ddl.NewOwnerManager(do.etcdClient, statistics.StatsPrompt, id, statistics.StatsOwnerKey, cancelFunc)
+		}
+		err := statsOwner.CampaignOwner(cancelCtx)
+		if err != nil {
```
Comment (zimulala) [inline domain/domain.go]: If we campaign failed, it may be not an owner always. Do we need to do something to handle it?

File: statistics/update.go [hunk 12]
```diff
@@ -133,3 +136,60 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) err
...
+	}
+	for _, idx := range tbl.Indices {
+		if len(idx.Buckets) > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// HandleAutoAnalyze analyze the newly created table or index.
```
Comment (zimulala) [inline statistics/update.go]: s/analyze/analyzes

File: statistics/update.go [hunk 13]
```diff
@@ -133,3 +136,60 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) err
...
+			statsTbl := h.GetTableStats(tbl.ID)
+			if statsTbl.Pseudo || statsTbl.Count == 0 {
+				continue
+			}
+			tblName := db.Name.O + "." + tbl.Name.O
+			if needAnalyzeTable(statsTbl, 20*h.Lease) {
+				sql := fmt.Sprintf("analyze table %s", tblName)
+				log.Infof("[stats] auto analyze table %s now", tblName)
+				_, _, err := h.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(h.ctx, sql)
+				return errors.Trace(err)
```
Comment (zimulala) [inline statistics/update.go]: Do we need to analyze the next table? Here we only print the log?

Comment (lamxTyler) [inline statistics/update.go]: 1. Nope, currently every time we only analyze one table or index, to reduce the impact on the server. 
2. Only the log? It also analyzes the table.

File: domain/domain.go (hunk 11 above)
Comment (lamxTyler) [inline domain/domain.go]: As long as there is at most one owner, it will be ok.

File: domain/domain.go (hunk 10 above)
Comment (lamxTyler) [inline domain/domain.go]: ok.

File: ddl/owner_manager.go (hunk 7 above)
Comment (lamxTyler) [inline ddl/owner_manager.go]: I search the project and find that no one uses it. Need to add back?

File: statistics/update.go (hunk 13 above)
Comment (zimulala) [inline statistics/update.go]: About the first point of view，you use `db.Tables` and haven't a `break` after handling one table or index, so how do we do it?

File: domain/domain.go (hunk 11 above)
Comment (zimulala) [inline domain/domain.go]: If all TiDB are failed, then what should we do？

File: statistics/update.go (hunk 13 above)
Comment (lamxTyler) [inline statistics/update.go]: There is a `return` after the execute?

File: domain/domain.go (hunk 11 above)
Comment (lamxTyler) [inline domain/domain.go]: Then there will be no auto analyze since there is no owner. Or do we have better solutions?

**Comment by hanfei1991** [comment]
@lamxTyler I think owner is not a kind of "util". We'd better isolate this package.

File: ddl/syncer.go [hunk 14]
```diff
@@ -34,6 +35,9 @@ const (
 	// DDLGlobalSchemaVersion is the path on etcd that is used to store the latest schema versions.
 	// It's exported for testing.
 	DDLGlobalSchemaVersion = "/tidb/ddl/global_schema_version"
+	// DDLOwnerKey is the ddl owner path that is saved to etcd, and it's exported for testing.
+	DDLOwnerKey = "/tidb/ddl/fg/owner"
+	ddlPrompt   = "ddl"
```
Comment (zimulala) [inline ddl/syncer.go]: It's better to put this code into `ddl.go`.

File: owner/manager.go [hunk 15]
```diff
@@ -32,88 +30,79 @@ import (
 	"google.golang.org/grpc"
 )
 
-// OwnerManager is used to campaign the owner and manage the owner information.
-type OwnerManager interface {
-	// ID returns the ID of DDL.
+// Manager is used to campaign the owner and manage the owner information.
```
Comment (zimulala) [inline owner/manager.go]: Why change this name?

File: statistics/update.go (hunk 13 above)
Comment (zimulala) [inline statistics/update.go]: Gotcha.

File: owner/manager.go (hunk 15 above)
Comment (lamxTyler) [inline owner/manager.go]: Because the package's name is `owner`, so it will make golint unhappy if the names' prefix is `owner`.

File: statistics/update.go [hunk 16]
```diff
@@ -133,3 +137,62 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) err
...
+
+// HandleAutoAnalyze analyzes the newly created table or index.
+func (h *Handle) HandleAutoAnalyze(is infoschema.InfoSchema) error {
+	dbs := is.AllSchemaNames()
+	for _, db := range dbs {
+		tbls := is.SchemaTables(model.NewCIStr(db))
+		for _, tbl := range tbls {
+			tblInfo := tbl.Meta()
+			statsTbl := h.GetTableStats(tblInfo.ID)
+			if statsTbl.Pseudo || statsTbl.Count == 0 {
```
Comment (hanfei1991) [inline statistics/update.go]: 1. We should analyze table with least version firstly. We can use a 'select table_id from stats_meta where modify_count >= count and count >= 1000 order by version limit 1.
2. If a table is deleted, it's count is not zero but histogram is empty. We shouldn't analyze it. If its version is old enough, we can delete it.

Comment (lamxTyler) [inline statistics/update.go]: 1. First, There is no direct mapping from table_id to table_name, which would make it hard to produce the analyze statement. Second, modify_count >= count does not ensure that this is a newly created table. Third, Since I will implement a job queue, the job queue could be a priority queue, which I think is a better solution.
2. Well, I did not notice it before. A deleted table has non-zero count is reasonable. So, let's directly make the deleted table's count 0.

Comment (hanfei1991) [inline statistics/update.go]: you can refine 1 in the future. And currently the deleted table is not with count 0, so we still need to process this case. Owner do the grabage collection is also nessesary.

Comment (lamxTyler) [inline statistics/update.go]: Make the deleted table 0 could be done in this PR. Garbage collection is not relevant with auto analyze, so I will do it in another PR.

Comment (lamxTyler) [inline statistics/update.go]: Also, even if the deleted table's count is not 0, we do not need to process it, because if it is deleted, then it will not be in the info schema or the stats cache.

**Comment by hanfei1991** [comment]
LGTM

File: owner/manager.go [hunk 17]
```diff
@@ -256,9 +240,11 @@ func (m *ownerManager) watchOwner(ctx goctx.Context, etcdSession *concurrency.Se
 	}
 }
 
-func init() {
-	err := setManagerSessionTTL()
```
Comment (zimulala) [inline owner/manager.go]: Why remove this function？ It's useful.

Comment (lamxTyler) [inline owner/manager.go]: But it is not used?

Comment (zimulala) [inline owner/manager.go]: It's used in the test.

File: domain/domain.go [hunk 18]
```diff
@@ -546,41 +548,69 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
...
+func (do *Domain) autoAnalyzeWorker(lease time.Duration) {
+	id := do.ddl.OwnerManager().ID()
+	cancelCtx, cancelFunc := goctx.WithCancel(goctx.Background())
+	var statsOwner owner.Manager
+	if do.etcdClient == nil {
+		statsOwner = owner.NewMockManager(id, cancelFunc)
+	} else {
+		statsOwner = owner.NewOwnerManager(do.etcdClient, statistics.StatsPrompt, id, statistics.StatsOwnerKey, cancelFunc)
+	}
+	err := statsOwner.CampaignOwner(cancelCtx)
```
Comment (zimulala) [inline domain/domain.go]: I think we need to do something when `err != nil`.

File: statistics/update_test.go [hunk 19]
```diff
@@ -223,3 +223,50 @@ func (s *testStatsUpdateSuite) TestTxnWithFailure(c *C) {
 	stats1 = h.GetTableStats(tableInfo1.ID)
 	c.Assert(stats1.Count, Equals, int64(rowCount1+1))
 }
+
+func (s *testStatsUpdateSuite) TestAutoUpdate(c *C) {
+	store, do, err := newStoreWithBootstrap()
```
Comment (zimulala) [inline statistics/update_test.go]: Can we use a unified store in testStatsUpdateSuite? Bootstrap make the test run a long time.

File: domain/domain.go (hunk 18 above)
Comment (lamxTyler) [inline domain/domain.go]: Like what?

File: statistics/update_test.go (hunk 19 above)
Comment (lamxTyler) [inline statistics/update_test.go]: Do it in another pr? There are lots of bootstrap in this package.

File: owner/manager.go [hunk 20]
```diff
@@ -259,6 +259,15 @@ func (m *ownerManager) watchOwner(ctx goctx.Context, etcdSession *concurrency.Se
 func init() {
 	err := setManagerSessionTTL()
 	if err != nil {
-		log.Warnf("[ddl] set manager session TTL failed %v", err)
+		log.Warnf("set manager session TTL failed %v", err)
+	}
+}
+
+func isContextDone(ctx goctx.Context) bool {
```
Comment (zimulala) [inline owner/manager.go]: Is this function useless?

File: domain/domain.go (hunk 18 above)
Comment (zimulala) [inline domain/domain.go]: The network between PD and TiDB is unstable

File: statistics/update_test.go (hunk 19 above)
Comment (zimulala) [inline statistics/update_test.go]: OK

File: owner/manager.go (hunk 20 above)
Comment (lamxTyler) [inline owner/manager.go]: Nope. In fact, this function also exists in DDL package, but it seems awkward to export it here. Any advice?

File: domain/domain.go (hunk 18 above)
Comment (lamxTyler) [inline domain/domain.go]: Try until the `err` is not nil?

Comment (zimulala) [inline domain/domain.go]: It‘s better to put it out of a goroutine. If it hard to handle, you can add a `TODO` here.

File: owner/manager.go (hunk 20 above)
Comment (zimulala) [inline owner/manager.go]: It's OK now.

File: domain/domain.go (hunk 18 above)
Comment (lamxTyler) [inline domain/domain.go]: OK, I add a TODO here.

**Comment by zimulala** [comment]
LGTM
Please resolve the conflict. @lamxTyler

**Comment by lamxTyler** [comment]
@zimulala  Resolved.
